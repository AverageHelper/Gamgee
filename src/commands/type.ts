import type { Command } from "./Command";
import { deleteMessage } from "../actions/messages/deleteMessage";

const type: Command = {
  name: "t",
  description: "Start typing. :wink:",
  async execute({ message, client, logger }) {
    const channel = message.channel;

    logger.debug(
      `${client.user?.username.concat(" is") ?? "I am"} typing in channel ${channel.id}...`
    );
    await deleteMessage(message, "Spam: Users don't need to see this command run");

    setTimeout(() => {
      channel.stopTyping(true);
    }, 5000);

    await channel.startTyping();
    logger.debug(`Finished typing in channel ${channel.id}`);
  }
};

export default type;
