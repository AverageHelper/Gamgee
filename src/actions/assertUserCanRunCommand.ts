import type { Command, CommandPermission, Subcommand } from "../commands/index.js";
import type { GuildMember, GuildTextBasedChannel } from "discord.js";
import { ApplicationCommandPermissionType } from "discord.js";
import { assertUnreachable } from "../helpers/assertUnreachable.js";
import { isPermissionAliasList, resolvePermissions } from "../commands/CommandPermission.js";
import { useLogger } from "../logger.js";
import {
	userHasPermissionInChannel,
	userHasRoleInGuild,
} from "../permissions/userHasOneOfRoles.js";

export type Invocable = Command | Subcommand;

const logger = useLogger();

/**
 * Assesses whether the calling guild member is allowed to run the given command.
 *
 * @param member The calling guild member.
 * @param command The command the user wants to run.
 * @param channel The channel in which the command is being run.
 *
 * @returns `true` if the user may be permitted to run the command. `false` otherwise.
 */
export async function assertUserCanRunCommand(
	member: GuildMember,
	command: Invocable,
	channel: GuildTextBasedChannel | null,
): Promise<boolean> {
	if (command.requiresGuild && !channel) {
		logger.debug(`Command '${command.name}' reqires a guild, but we don't have one right now.`);
		return false;
	}

	if (!command.permissions) {
		// No permissions demanded
		logger.debug(`Command '${command.name}' does not require specific user permissions.`);
		logger.debug("Proceeding...");
		return true;
	}

	const guild = channel?.guild;
	const permissions: ReadonlyArray<CommandPermission> = guild
		? isPermissionAliasList(command.permissions)
			? await resolvePermissions(command.permissions, guild)
			: await command.permissions(guild)
		: [];

	logger.debug(
		`Command '${command.name}' requires that callers satisfy 1 of ${permissions.length} cases:`,
	);

	let idx = 0;
	for (const access of permissions) {
		idx += 1;
		switch (access.type) {
			case ApplicationCommandPermissionType.Role: {
				logger.debug(
					`\tCase ${idx}: User must${access.permission ? "" : " not"} have ROLE ID: ${access.id}`,
				);
				const userHasRole =
					guild !== undefined && (await userHasRoleInGuild(member, access.id, guild));
				logger.debug(`\tUser ${userHasRole ? "has" : "does not have"} role ${access.id}`);
				if (access.permission && userHasRole) {
					logger.debug("\tProceeding...");
					return true;
				}
				break;
			}

			case ApplicationCommandPermissionType.User: {
				logger.debug(
					`\tCase ${idx}: User must${access.permission ? "" : " not"} have USER ID: ${access.id}`,
				);
				const userHasId = member.id === access.id;
				logger.debug(`\tUser ${userHasId ? "has" : "does not have"} ID ${access.id}`);
				if (access.permission && userHasId) {
					logger.debug("\tProceeding...");
					return true;
				}
				break;
			}

			case ApplicationCommandPermissionType.Channel: {
				logger.debug(
					`\tCase ${idx}: User must${
						access.permission ? "" : " not"
					} have 'ReadMessageHistory' access in CHANNEL ID: ${access.id}`,
				);
				const userCanSeeChannel =
					channel !== null && userHasPermissionInChannel(member, "ReadMessageHistory", access.id);
				logger.debug(
					`\tUser ${
						userCanSeeChannel ? "has" : "does not have"
					} permission to read messages in channel ${access.id}`,
				);
				if (access.permission && channel) {
					logger.debug("\tProceeding...");
					return true;
				}
				break;
			}

			default:
				return assertUnreachable(access.type);
		}
	}

	// Caller fails permissions checks
	logger.debug("User did not pass any permission checks.");
	return false;
}
