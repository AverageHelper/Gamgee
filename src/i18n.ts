/* eslint-disable @typescript-eslint/no-var-requires */

// ** Install language files here **
const vocabulary = {
	de: require("./locales/de.json") as typeof import("./locales/de.json"),
	"en-GB": require("./locales/en-GB.json") as typeof import("./locales/en-GB.json"),
	"en-US": require("./locales/en-US.json") as typeof import("./locales/en-US.json"),
	"es-ES": require("./locales/es-ES.json") as typeof import("./locales/es-ES.json"),
	fr: require("./locales/fr.json") as typeof import("./locales/fr.json"),
	hu: require("./locales/hu.json") as typeof import("./locales/hu.json"),
	"pt-BR": require("./locales/pt-BR.json") as typeof import("./locales/pt-BR.json")
} as const;

/* eslint-disable @typescript-eslint/no-var-requires */

const DEFAULT_LOCALE = "en-US";

export type SupportedLocale = keyof typeof vocabulary;

export const locales = Object.keys(vocabulary) as ReadonlyArray<SupportedLocale>;

/** Returns `true` if the given string matches a supported locale code. */
export function isSupportedLocale(tbd: unknown): tbd is SupportedLocale {
	return locales.includes(tbd as SupportedLocale);
}

// TODO: Validate that all of our strings files match the master schema
// TypeScript ensures here that DEFAULT_LOCALE is a valid locale:
type MessageSchema = typeof vocabulary[typeof DEFAULT_LOCALE];

// ** I18N Utilities **

interface RecursiveStringRecord extends Record<string, string | RecursiveStringRecord> {}

// for testing i18n over generic vocabularies
type Vocabulary = {
	[DEFAULT_LOCALE]: RecursiveStringRecord;
} & {
	[P in Exclude<SupportedLocale, typeof DEFAULT_LOCALE>]?: RecursiveStringRecord;
};

// eslint-disable-next-line @typescript-eslint/no-unused-vars
const _: Vocabulary = vocabulary; // ensures our language types are formatted right

import type { Get, Split } from "type-fest";
import _get from "lodash/get.js";
import isString from "lodash/isString.js";

function split<S extends string, D extends string>(string: S, separator: D): Split<S, D> {
	return string.split(separator) as Split<S, D>;
}

const get = <BaseType, Path extends string | ReadonlyArray<string>>(
	object: BaseType,
	path: Path
): Get<BaseType, Path> => _get(object, path) as Get<BaseType, Path>;

const DOT = ".";

/**
 * Returns the given `locale`'s translation for the given `keypath`.
 *
 * If the given `locale` does not contain a matching translation,
 * but the default locale does, then that translation is returned.
 *
 * Returns `keypath` if no matching translation exists.
 */
export function t<K extends string>(
	keypath: K,
	locale: SupportedLocale
): Get<MessageSchema, K> extends string ? Get<MessageSchema, K> : undefined;

/**
 * Returns the given `locale`'s translation for the given `keypath`.
 *
 * If the given `locale` does not contain a matching translation,
 * but the default locale does, then that translation is returned.
 *
 * Returns `keypath` if no matching translation exists.
 */
export function t<K extends string, V extends Vocabulary>(
	keypath: K,
	locale: SupportedLocale,
	messages: V
): Get<V[typeof DEFAULT_LOCALE], K> extends string ? Get<V[typeof DEFAULT_LOCALE], K> : undefined;

// FIXME: We shouldn't need to overload this
export function t<K extends string>(
	keypath: K,
	locale: SupportedLocale,
	data = vocabulary
): string | undefined {
	if (keypath === "") return undefined;

	const result = get(data[locale], split(keypath, DOT));

	if (isString(result) && result) return result; // found a result in the given locale!

	if (locale !== DEFAULT_LOCALE) {
		// recurse, try the default locale:
		return t(keypath, DEFAULT_LOCALE, data);
	}

	return undefined; // we're stumped, return nothing
}

/**
 * Returns an object with every translation we have for the given `keypath`,
 * or `undefined` if no translations exist.
 */
export function localizations<K extends string>(
	keypath: K
): Get<MessageSchema, K> extends string ? Partial<Record<SupportedLocale, string>> : undefined;

// FIXME: We shouldn't need to overload this
export function localizations<K extends string>(
	keypath: K
): Partial<Record<SupportedLocale, string>> | undefined {
	// Get all localizations for the given keypath
	const result: Partial<Record<SupportedLocale, string>> = {};

	locales.forEach(locale => {
		const translation = t(keypath, locale);
		if (translation !== undefined) {
			// only add the translation if there is one to add
			result[locale] = translation;
		}
	});

	// Return undefined if there were no results
	if (Object.keys(result).length === 0) return undefined;
	return result;
}

// For science:
/* eslint-disable @typescript-eslint/no-unused-vars */
// const enUS = vocabulary["en-US"];

/*
import type { Join } from "type-fest";

// We start with this at dev time, typing into a function:
const given = "commands.sr.name";

// Goal: Confirm that a given string makes up a valid key-path in a given object.

// Split given by DOT:
const p: ["commands", "sr", "name"] = split(given, DOT);

// const testBad: string = get(enUS, "commands.sr."); // Error
const testGood: string = get(enUS, "commands.sr.name");
const testAlsoBad: string = get(enUS, ["commands", "sr"]);
const testAlsoGood: string = get(enUS, p);

// If we can split `given` by a DOT "." we get these values:
const p0: "commands" = p[0];
const p1: "sr" = p[1];
const p2: "name" = p[2];

// We can recombine segments with this:
type Path<L extends string, R extends string, Sep extends string = typeof DOT> = Join<[L, R], Sep>;

const partial: Path<typeof p0, typeof p1> = `${p0}.${p1}`;

// these values are equivalent to `given`:
const path: Path<Path<typeof p0, typeof p1>, typeof p2> = given;
const alsoPath: Join<typeof p, typeof DOT> = given;

const commands: MessageSchema[typeof p0] = vocabulary["en-US"][p0];
const sr: MessageSchema[typeof p0][typeof p1] = vocabulary["en-US"][p0][p1];
const name: MessageSchema[typeof p0][typeof p1][typeof p2] = t(given, "en-US");

type IsKeyPath<O, K extends string> = Get<O, K> extends string ? true : false;

const good: IsKeyPath<MessageSchema, typeof given> = true;
const bad: IsKeyPath<MessageSchema, "typeof given"> = false;
*/
