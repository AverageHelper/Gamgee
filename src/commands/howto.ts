import type { GuildedCommand } from "./Command.js";
import { composed, createPartialString, push, pushNewLine } from "../helpers/composeStrings.js";
import { getCommandPrefix } from "../useGuildStorage.js";
import { localizations, ti } from "../i18n.js";
import { mentionCommand } from "../helpers/mentionCommands.js";
import { supportedPlatformsList } from "../constants/repository.js";

export const howto: GuildedCommand = {
	name: "howto",
	nameLocalizations: localizations("commands.howto.name"),
	description: "Print instructions for using the common queue commands.",
	descriptionLocalizations: localizations("commands.howto.description"),
	requiresGuild: true,
	async execute({ guildLocale, guild, type, reply }) {
		const { sr } = await import("./songRequest.js");
		const { nowPlaying } = await import("./nowPlaying.js");

		const srCommandName: string = sr.nameLocalizations
			? sr.nameLocalizations[guildLocale] ?? sr.name
			: sr.name;

		// Print the standard help
		const COMMAND_PREFIX = await getCommandPrefix(guild);
		const msg = createPartialString();

		const exampleQuery = "https://youtu.be/dQw4w9WgXcQ"; // :P

		const srMention = mentionCommand(sr, guild, COMMAND_PREFIX);
		push(
			ti(
				"commands.howto.responses.to-submit",
				{
					command: srMention,
					example: `\`/${srCommandName} ${exampleQuery}\``
				},
				guildLocale
			),
			msg
		);
		pushNewLine(msg);

		const supportedPlatforms =
			type === "interaction"
				? ti(
						"commands.howto.responses.see-supported-platforms",
						{ url: `<${supportedPlatformsList}>` },
						guildLocale
				  )
				: ti(
						"commands.howto.responses.see-supported-platforms-at-url",
						{ url: `<${supportedPlatformsList}>` },
						guildLocale
				  );
		push(supportedPlatforms, msg);
		pushNewLine(msg);
		pushNewLine(msg);

		const npMention = mentionCommand(nowPlaying, guild, COMMAND_PREFIX);
		push(
			ti("commands.howto.responses.to-get-current-song", { command: npMention }, guildLocale),
			msg
		);

		return await reply(composed(msg));
	}
};
