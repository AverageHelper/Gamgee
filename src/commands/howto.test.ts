import "../../tests/testUtils/leakedHandles.js";

jest.mock("../actions/assertUserCanRunCommand");

import { assertUserCanRunCommand } from "../actions/assertUserCanRunCommand.js";
const mockAssertUserCanRunCommand = assertUserCanRunCommand as jest.Mock;

import type { GuildedCommandContext } from "./Command.js";
import { howto } from "./howto.js";
import { locales } from "../i18n.js";

const mockReply = jest.fn();

describe.each(locales)("How-to command in %s", guildLocale => {
	let context: GuildedCommandContext;

	beforeEach(() => {
		context = {
			guild: { id: "some-guild" },
			guildLocale,
			reply: mockReply,
			type: "message"
		} as unknown as GuildedCommandContext;

		mockAssertUserCanRunCommand.mockResolvedValue(true);
	});

	test("informs the user how to run queue commands (message)", async () => {
		await howto.execute(context);
		expect(mockReply).toHaveBeenCalledOnce();
		expect(mockReply).toHaveBeenCalledWith(expect.toBeString());

		const calls = mockReply.mock.calls[0] as Array<unknown>;
		const description = calls[0];
		expect(description).toMatchSnapshot();
	});

	test("informs the user how to run queue commands (interaction)", async () => {
		context = { ...context, type: "interaction" } as unknown as GuildedCommandContext;
		await howto.execute(context);
		expect(mockReply).toHaveBeenCalledOnce();
		expect(mockReply).toHaveBeenCalledWith(expect.toBeString());

		const calls = mockReply.mock.calls[0] as Array<unknown>;
		const description = calls[0];
		expect(description).toMatchSnapshot();
	});
});
