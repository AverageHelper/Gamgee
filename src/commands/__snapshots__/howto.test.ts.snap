// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`How-to command in de > informs the user how to run queue commands (interaction) 1`] = `
"Um einen Song einzureichen, verwende \`?sr\`

Zum Beispiel: \`/sr https://youtu.be/dQw4w9WgXcQ\`
Ich werde mit einer Bestätigung antworten, dass dein Song zur Warteschlange hinzugefügt wurde!
Siehe [unsere Liste der unterstützten Plattformen](<https://github.com/AverageHelper/Gamgee#supported-music-platforms>).

Um einen Link zum aktuellen Song zu erhalten, verwende \`?nowplaying\`."
`;

exports[`How-to command in de > informs the user how to run queue commands (message) 1`] = `
"Um einen Song einzureichen, verwende \`?sr\`

Zum Beispiel: \`/sr https://youtu.be/dQw4w9WgXcQ\`
Ich werde mit einer Bestätigung antworten, dass dein Song zur Warteschlange hinzugefügt wurde!
Siehe unsere Liste der unterstützten Plattformen <https://github.com/AverageHelper/Gamgee#supported-music-platforms>.

Um einen Link zum aktuellen Song zu erhalten, verwende \`?nowplaying\`."
`;

exports[`How-to command in en-GB > informs the user how to run queue commands (interaction) 1`] = `
"To submit a song, use \`?sr\`

For example: \`/sr https://youtu.be/dQw4w9WgXcQ\`
I will respond with confirmation that your song has joined the queue!
See [our list of supported platforms](<https://github.com/AverageHelper/Gamgee#supported-music-platforms>).

To get a link to the current song, use \`?nowplaying\`."
`;

exports[`How-to command in en-GB > informs the user how to run queue commands (message) 1`] = `
"To submit a song, use \`?sr\`

For example: \`/sr https://youtu.be/dQw4w9WgXcQ\`
I will respond with confirmation that your song has joined the queue!
See our list of supported platforms at <https://github.com/AverageHelper/Gamgee#supported-music-platforms>.

To get a link to the current song, use \`?nowplaying\`."
`;

exports[`How-to command in en-US > informs the user how to run queue commands (interaction) 1`] = `
"To submit a song, use \`?sr\`

For example: \`/sr https://youtu.be/dQw4w9WgXcQ\`
I will respond with confirmation that your song has joined the queue!
See [our list of supported platforms](<https://github.com/AverageHelper/Gamgee#supported-music-platforms>).

To get a link to the current song, use \`?nowplaying\`."
`;

exports[`How-to command in en-US > informs the user how to run queue commands (message) 1`] = `
"To submit a song, use \`?sr\`

For example: \`/sr https://youtu.be/dQw4w9WgXcQ\`
I will respond with confirmation that your song has joined the queue!
See our list of supported platforms at <https://github.com/AverageHelper/Gamgee#supported-music-platforms>.

To get a link to the current song, use \`?nowplaying\`."
`;

exports[`How-to command in es-ES > informs the user how to run queue commands (interaction) 1`] = `
"Para enviar una canción, use \`?sr\`

Por ejemplo: \`/sr https://youtu.be/dQw4w9WgXcQ\`
¡Responderé con una confirmación indicando que tu canción se ha unido a la cola!
Vea [nuestra lista de plataformas compatibles](<https://github.com/AverageHelper/Gamgee#supported-music-platforms>).

Para obtener un enlace a la canción actual, use \`?nowplaying\`."
`;

exports[`How-to command in es-ES > informs the user how to run queue commands (message) 1`] = `
"Para enviar una canción, use \`?sr\`

Por ejemplo: \`/sr https://youtu.be/dQw4w9WgXcQ\`
¡Responderé con una confirmación indicando que tu canción se ha unido a la cola!
Consulte nuestra lista de plataformas compatibles en <https://github.com/AverageHelper/Gamgee#supported-music-platforms>.

Para obtener un enlace a la canción actual, use \`?nowplaying\`."
`;

exports[`How-to command in fr > informs the user how to run queue commands (interaction) 1`] = `
"Pour soumettre une chanson, utilisez \`?sr\`

Par exemple: \`/sr https://youtu.be/dQw4w9WgXcQ\`
Je vous répondrai avec une confirmation indiquant que votre chanson a rejoint la file d'attente!
Consultez [notre liste de plateformes prises en charge](<https://github.com/AverageHelper/Gamgee#supported-music-platforms>).

Pour obtenir un lien vers la chanson actuelle, utilisez \`?nowplaying\`."
`;

exports[`How-to command in fr > informs the user how to run queue commands (message) 1`] = `
"Pour soumettre une chanson, utilisez \`?sr\`

Par exemple: \`/sr https://youtu.be/dQw4w9WgXcQ\`
Je vous répondrai avec une confirmation indiquant que votre chanson a rejoint la file d'attente!
Consultez notre liste des plates-formes prises en charge sur <https://github.com/AverageHelper/Gamgee#supported-music-platforms>.

Pour obtenir un lien vers la chanson actuelle, utilisez \`?nowplaying\`."
`;

exports[`How-to command in hu > informs the user how to run queue commands (interaction) 1`] = `
"Dal beküldéséhez használja a \`?sr\` parancsot

Például: \`/sr https://youtu.be/dQw4w9WgXcQ\`
Jelezni fogom, amint sikeresen hozzáadtam a dalt a sorhoz!
Tekintse meg [a támogatott platformok listáját](<https://github.com/AverageHelper/Gamgee#supported-music-platforms>).

Az aktuális dal hivatkozásának lekéréséhez használja a \`?nowplaying\` parancsot."
`;

exports[`How-to command in hu > informs the user how to run queue commands (message) 1`] = `
"Dal beküldéséhez használja a \`?sr\` parancsot

Például: \`/sr https://youtu.be/dQw4w9WgXcQ\`
Jelezni fogom, amint sikeresen hozzáadtam a dalt a sorhoz!
Tekintse meg a támogatott platformok listáját a <https://github.com/AverageHelper/Gamgee#supported-music-platforms> címen.

Az aktuális dal hivatkozásának lekéréséhez használja a \`?nowplaying\` parancsot."
`;

exports[`How-to command in pt-BR > informs the user how to run queue commands (interaction) 1`] = `
"Para enviar uma música, use \`?sr\`

Por exemplo: \`/sr https://youtu.be/dQw4w9WgXcQ\`
Responderei com uma confirmação indicando que sua música entrou na fila!
Veja [nossa lista de plataformas suportadas](<https://github.com/AverageHelper/Gamgee#supported-music-platforms>).

Para obter um link para a música atual, use \`?nowplaying\`."
`;

exports[`How-to command in pt-BR > informs the user how to run queue commands (message) 1`] = `
"Para enviar uma música, use \`?sr\`

Por exemplo: \`/sr https://youtu.be/dQw4w9WgXcQ\`
Responderei com uma confirmação indicando que sua música entrou na fila!
Veja nossa lista de plataformas suportadas em <https://github.com/AverageHelper/Gamgee#supported-music-platforms>.

Para obter um link para a música atual, use \`?nowplaying\`."
`;
